### PR TITLE
Atomize Rename operation when encryption is enabled

### PR DESCRIPTION
**Pull request by hunterlxt**
https://github.com/tikv/rocksdb/pull/222 used `LinkFile` instead of `RenameFile` api of key manager. But `LinkFile` needs check the dst file information, in `RenameFile` logic, we don't care about that. So just skip encryption for current file.

Signed-off-by: Xintao <hunterlxt@live.com>

### CODE REVIEW

**Review by Connor1996 — APPROVED** [review]

**Review by tabokie — COMMENTED** [review]
This case is covered by https://github.com/tikv/rocksdb/blob/2dfbfa51b7892a80debc18e52c863b93bd79e491/env/env_basic_test.cc#L217-L220 we should add an option to test encrypted env.

**Comment by hunterlxt** [comment]
> This case is covered by
> 
> https://github.com/tikv/rocksdb/blob/2dfbfa51b7892a80debc18e52c863b93bd79e491/env/env_basic_test.cc#L217-L220
> 
> we should add an option to test encrypted env.

Will the CI test not include it?  If not, how to turn it on. Sorry I'm not familiar with rocksdb tests

**Comment by tabokie** [comment]
@hunterlxt like this snippet: https://github.com/tikv/rocksdb/blob/2dfbfa51b7892a80debc18e52c863b93bd79e491/env/env_basic_test.cc#L127-L131
But use TestKeyManager and KeyManagedEncryptedEnv instead and wrap it with LITE and OPENSSL macro.

File: encryption/encryption.cc [hunk 1]
```diff
@@ -440,6 +440,12 @@ Status KeyManagedEncryptedEnv::LinkFile(const std::string& src_fname,
 
 Status KeyManagedEncryptedEnv::RenameFile(const std::string& src_fname,
                                           const std::string& dst_fname) {
+  // In order to ensure that the key manager does not have this dst_fname
+  // information, it may exist in the file system, rename will rewrite
+  // the file, so this is acceptable.
+  Status delete_status __attribute__((__unused__)) =
```
Comment (yiwu-arbug) [inline encryption/encryption.cc:446]: should return error if this call fail.

File: env/env_basic_test.cc [hunk 2]
```diff
@@ -124,11 +124,21 @@ class DummyFileSystemInspector : public FileSystemInspector {
   size_t refill_bytes_;
 };
 
-static std::unique_ptr<Env> inspected_env(
+static std::unique_ptr<Env> inspected_fs_env(
```
Comment (tabokie) [inline env/env_basic_test.cc:127]:
```suggestion
static std::unique_ptr<Env> fs_inspected_env(
```

File: env/env_basic_test.cc [hunk 3]
```diff
@@ -124,11 +124,21 @@ class DummyFileSystemInspector : public FileSystemInspector {
...
     NewFileSystemInspectedEnv(new NormalizingEnvWrapper(Env::Default()),
                               std::make_shared<DummyFileSystemInspector>(1)));
 INSTANTIATE_TEST_CASE_P(FileSystemInspectedEnv, EnvBasicTestWithParam,
-                        ::testing::Values(inspected_env.get()));
+                        ::testing::Values(inspected_fs_env.get()));
+
+#ifdef ROCKSDB_LITE
+#ifdef OPENSSL
+static std::unique_ptr<Env> inspected_encrypt_env(
```
Comment (tabokie) [inline env/env_basic_test.cc:135]:
```suggestion
static std::unique_ptr<Env> key_managed_encrypted_env(
```

File: encryption/encryption.cc [hunk 4]
```diff
@@ -440,9 +440,16 @@ Status KeyManagedEncryptedEnv::LinkFile(const std::string& src_fname,
 
 Status KeyManagedEncryptedEnv::RenameFile(const std::string& src_fname,
                                           const std::string& dst_fname) {
+  // In order to ensure that the key manager does not have this dst_fname
+  // information, it may exist in the file system, rename will rewrite
+  // the file, so this is acceptable.
+  Status s = = key_manager_->DeleteFile(dst_fname);
```
Comment (yiwu-arbug) [inline encryption/encryption.cc:446]:
```suggestion
  Status s = key_manager_->DeleteFile(dst_fname);
```

File: env/env_basic_test.cc [hunk 5]
```diff
@@ -124,11 +124,21 @@ class DummyFileSystemInspector : public FileSystemInspector {
...
 
-static std::unique_ptr<Env> inspected_env(
+static std::unique_ptr<Env> inspected_fs_env(
     NewFileSystemInspectedEnv(new NormalizingEnvWrapper(Env::Default()),
                               std::make_shared<DummyFileSystemInspector>(1)));
 INSTANTIATE_TEST_CASE_P(FileSystemInspectedEnv, EnvBasicTestWithParam,
-                        ::testing::Values(inspected_env.get()));
+                        ::testing::Values(inspected_fs_env.get()));
+
+#ifdef ROCKSDB_LITE
```
Comment (tabokie) [inline env/env_basic_test.cc:133]: Seems we already have a LITE protection below, can remove this.

File: env/env_basic_test.cc [hunk 6]
```diff
@@ -124,11 +124,21 @@ class DummyFileSystemInspector : public FileSystemInspector {
...
+#ifdef OPENSSL
+static std::unique_ptr<Env> inspected_encrypt_env(
+    NewKeyManagedEncryptedEnv(new NormalizingEnvWrapper(Env::Default()),
+                              std::make_shared<KeyManager>(TestKeyManager)));
+INSTANTIATE_TEST_CASE_P(FileSystemInspectedEnv, EnvBasicTestWithParam,
+                        ::testing::Values(inspected_encrypt_env.get()));
+#endif
```
Comment (tabokie) [inline env/env_basic_test.cc:140]:
```suggestion
#endif  // OPENSSL
```

**Review by tabokie — APPROVED** [review]
minor comments

**Comment by yiwu-arbug** [comment]
LGTM on my side. Leave to @tabokie for approval.

File: env/env_basic_test.cc [hunk 7]
```diff
@@ -124,11 +124,19 @@ class DummyFileSystemInspector : public FileSystemInspector {
...
     NewFileSystemInspectedEnv(new NormalizingEnvWrapper(Env::Default()),
                               std::make_shared<DummyFileSystemInspector>(1)));
 INSTANTIATE_TEST_CASE_P(FileSystemInspectedEnv, EnvBasicTestWithParam,
-                        ::testing::Values(inspected_env.get()));
+                        ::testing::Values(fs_inspected_env.get()));
+
+#ifdef OPENSSL
+static std::unique_ptr<Env> key_managed_encrypted_env(
+    NewKeyManagedEncryptedEnv(new NormalizingEnvWrapper(Env::Default()),
+                              std::make_shared<KeyManager>(TestKeyManager)));
```
Comment (tabokie) [inline env/env_basic_test.cc:136]: Build is failing. What is this thing? `TestKeyManager` refers to the test artifact in `db_test_util.h`.

File: env/env_basic_test.cc [hunk 8]
```diff
@@ -124,11 +124,19 @@ class DummyFileSystemInspector : public FileSystemInspector {
...
+INSTANTIATE_TEST_CASE_P(FileSystemInspectedEnv, EnvBasicTestWithParam,
```
Comment (tabokie) [inline env/env_basic_test.cc:137]: `FileSystemInspectedEnv` is the name used to generate corresponding test functions, in this case should rename to `KeyManagedEncryptedEnv`.

File: env/env_basic_test.cc [hunk 9]
```diff
@@ -124,14 +125,27 @@ class DummyFileSystemInspector : public FileSystemInspector {
...
 INSTANTIATE_TEST_CASE_P(FileSystemInspectedEnv, EnvBasicTestWithParam,
-                        ::testing::Values(inspected_env.get()));
+                        ::testing::Values(fs_inspected_env.get()));
+
+#ifdef OPENSSL
+const std::string TestKeyManager::default_key =
+    "\x12\x34\x56\x78\x12\x34\x56\x78\x12\x34\x56\x78\x12\x34\x56\x78\x12\x34"
+    "\x56\x78\x12\x34\x56\x78";
+const std::string TestKeyManager::default_iv =
+    "\xaa\xbb\xcc\xdd\xaa\xbb\xcc\xdd\xaa\xbb\xcc\xdd\xaa\xbb\xcc\xdd";
```
Comment (tabokie) [inline env/env_basic_test.cc:139]: It's utterly wrong to try hide a compile error with this.

**Review by tabokie — CHANGES_REQUESTED** [review]

File: env/env_basic_test.cc (hunk 9 above)
Comment (yiwu-arbug) [reply env/env_basic_test.cc:139]: move `TestKeyManager` from db/db_test_util.(cc|h) to test_util/testutil.(cc|h) may solve the issue.

File: test_util/testutil.h [hunk 10]
```diff
@@ -27,6 +29,39 @@
...
+    file_info->iv = default_iv;
+    return Status::OK();
+  }
+
+  Status DeleteFile(const std::string&) override { return Status::OK(); }
+  Status LinkFile(const std::string&, const std::string&) override {
+    return Status::OK();
+  }
+};
+#endif
```
Comment (tabokie) [inline test_util/testutil.h:63]: almost there, TestKeyManager needs to be wrapped in `ifndef LITE`.

File: encryption/encryption.cc [hunk 11]
```diff
@@ -15,6 +15,17 @@ namespace rocksdb {
 namespace encryption {
 
 namespace {
+// For some files, we don't require them to be encrypted.
```
Comment (tabokie) [inline encryption/encryption.cc:18]: New changes seem unrelated to this PR?

File: encryption/encryption.cc [hunk 12]
```diff
@@ -15,6 +15,17 @@ namespace rocksdb {
...
+bool skip_encryption(const std::string& fname) {
+  std::string skip_name = "CURRENT";
```
Comment (yiwu-arbug) [inline encryption/encryption.cc:20]: 1. Define a constant for the current file in filename/filename.h. Also update that file to use the constant.
2. Add comment why we want to exclude the current file from encryption.

File: encryption/encryption.cc (hunk 11 above)
Comment (yiwu-arbug) [reply encryption/encryption.cc:18]: @tabokie We find that rocksdb rely on `rename` being atomic to update the `CURRENT` file when creating a new manifest. For a quick fix we are to exclude the `CURRENT` file from being encrypted.

File: encryption/encryption.cc (hunk 12 above)
Comment (tabokie) [reply encryption/encryption.cc:20]: also kEncryptionSkipSuffix is a better name for it.

File: encryption/encryption.cc [hunk 13]
```diff
@@ -15,6 +15,17 @@ namespace rocksdb {
...
+  size_t fname_idx = fname.rfind(skip_name);
+  if (fname.length() - fname_idx != skip_name.length()) {
```
Comment (tabokie) [inline encryption/encryption.cc:22]:
```suggestion
  size_t check_length = skip_name.length();
  if (fname.length() >= check_length && !fname.compare(fname.length() - check_length, check_length, skip_name)) {
    return true;
  }
```

File: file/filename.cc [hunk 14]
```diff
@@ -24,6 +24,18 @@ namespace rocksdb {
 static const std::string kRocksDbTFileExt = "sst";
 static const std::string kLevelDbTFileExt = "ldb";
 static const std::string kRocksDBBlobFileExt = "blob";
+static const std::string kEncryptionSkipName  = "CURRENT";
+
+bool ShouldSkipEncryption(const std::string& fname) {
+  size_t check_length = kEncryptionSkipName.length();
```
Comment (tabokie) [inline file/filename.cc:30]: check equal against "CURRENT" (can be hardcoded) and check suffix against `kUnencryptedTempFileNameSuffix`

File: encryption/encryption.cc [hunk 15]
```diff
@@ -440,6 +452,14 @@ Status KeyManagedEncryptedEnv::LinkFile(const std::string& src_fname,
 
 Status KeyManagedEncryptedEnv::RenameFile(const std::string& src_fname,
                                           const std::string& dst_fname) {
+  if (ShouldSkipEncryption(dst_fname)) {
+    Status s = target()->RenameFile(src_fname, dst_fname);
+    if (!s.ok()) {
+      return s;
+    }
+    s = key_manager_->DeleteFile(src_fname);
```
Comment (tabokie) [inline encryption/encryption.cc:460]: No need

File: file/filename.h [hunk 16]
```diff
@@ -42,6 +42,9 @@ enum FileType {
   kBlobFile
 };
 
+// For some files, we don't require them to be encrypted.
```
Comment (tabokie) [inline file/filename.h:45]:
```suggestion
// Some non-sensitive files are not encrypted to preserve atomicity of file operations.
```

File: file/filename.cc [hunk 17]
```diff
@@ -364,7 +376,8 @@ Status SetCurrentFile(Env* env, const std::string& dbname,
   Slice contents = manifest;
   assert(contents.starts_with(dbname + "/"));
   contents.remove_prefix(dbname.size() + 1);
-  std::string tmp = TempFileName(dbname, descriptor_number);
+  std::string tmp =
+      TempFileName(dbname, descriptor_number) + "." + kEncryptionSkipName;
```
Comment (tabokie) [inline file/filename.cc:380]: Add a `kUnencryptedTempFileNameSuffix` and a method named `UnencryptedTempFileName`?

Comment (tabokie) [reply file/filename.cc:380]: `tmp.plain` sounds ok

File: encryption/encryption.cc (hunk 15 above)
Comment (tabokie) [reply encryption/encryption.cc:460]: Check that src_fname is also SkipEncryption, return an invalid argument error if not.

File: encryption/encryption.cc [hunk 18]
```diff
@@ -440,6 +452,14 @@ Status KeyManagedEncryptedEnv::LinkFile(const std::string& src_fname,
...
+    return s;
+  }
```
Comment (tabokie) [inline encryption/encryption.cc:462]: else assert(!ShouldSkipEncryption(src_fname))

File: encryption/encryption.cc [hunk 19]
```diff
@@ -425,6 +433,10 @@ Status KeyManagedEncryptedEnv::DeleteFile(const std::string& fname) {
 
 Status KeyManagedEncryptedEnv::LinkFile(const std::string& src_fname,
                                         const std::string& dst_fname) {
+  if (ShouldSkipEncryption(dst_fname)) {
```
Comment (tabokie) [inline encryption/encryption.cc:436]: same here

**Review by tabokie — APPROVED** [review]
